src/libponyc/reach/reach.c

```diff
@@ -1045,51 +1045,120 @@ static void reachable_ffi(reach_t* r, ast_t* ast, pass_opt_t* opt)
   }
 }
 
-static void reachable_identity(reach_t* r, ast_t* ast, pass_opt_t* opt)
+static void reachable_identity_type(reach_t* r, ast_t* l_type, ast_t* r_type,
+  pass_opt_t* opt)
 {
-  AST_GET_CHILDREN(ast, left, right);
+  if((ast_id(l_type) == TK_TUPLETYPE) && (ast_id(r_type) == TK_TUPLETYPE))
+  {
+    ast_t* l_child = ast_child(l_type);
+    ast_t* r_child = ast_child(r_type);
 
-  ast_t* l_type = ast_type(left);
-  ast_t* r_type = ast_type(right);
+    while(l_child != NULL)
+    {
+      reachable_identity_type(r, l_child, r_child, opt);
+      l_child = ast_sibling(l_child);
+      r_child = ast_sibling(r_child);
+    }
 
-  if(!is_known(l_type) && !is_known(r_type))
-  {
+    pony_assert(r_child == NULL);
+  } else if(!is_known(l_type) && !is_known(r_type)) {
     reach_type_t* r_left = reach_type(r, l_type);
     reach_type_t* r_right = reach_type(r, r_type);
+
     int sub_kind = subtype_kind_overlap(r_left, r_right);
 
     if((sub_kind & SUBTYPE_KIND_TUPLE) != 0)
     {
       const char* name = stringtab("__is");
 
+      if((reach_method_name(r_left, name) != NULL) &&
+        (reach_method_name(r_right, name) != NULL))
+        return;
+
       reach_method_name_t* n = add_method_name(r_left, name, true);
       add_rmethod(r, r_left, n, TK_BOX, NULL, opt, true);
 
       n = add_method_name(r_right, name, true);
       add_rmethod(r, r_right, n, TK_BOX, NULL, opt, true);
+
+      reach_type_t* l_sub;
+      size_t i = HASHMAP_BEGIN;
+
+      while((l_sub = reach_type_cache_next(&r_left->subtypes, &i)) != NULL)
+      {
+        if(l_sub->underlying != TK_TUPLETYPE)
+          continue;
+
+        reach_type_t k;
+        k.name = l_sub->name;
+        size_t j = HASHMAP_UNKNOWN;
+        reach_type_t* r_sub = reach_type_cache_get(&r_right->subtypes, &k, &j);
+
+        if(r_sub != NULL)
+        {
+          pony_assert(l_sub == r_sub);
+          reachable_identity_type(r, l_sub->ast_cap, r_sub->ast_cap, opt);
+        }
+      }
     }
   }
 }
 
-static void reachable_digestof(reach_t* r, ast_t* ast, pass_opt_t* opt)
+static void reachable_identity(reach_t* r, ast_t* ast, pass_opt_t* opt)
 {
-  ast_t* expr = ast_child(ast);
-  ast_t* type = ast_type(expr);
+  AST_GET_CHILDREN(ast, left, right);
+
+  ast_t* l_type = ast_type(left);
+  ast_t* r_type = ast_type(right);
+
+  reachable_identity_type(r, l_type, r_type, opt);
+}
 
-  if(!is_known(type))
+static void reachable_digestof_type(reach_t* r, ast_t* type, pass_opt_t* opt)
+{
+  if(ast_id(type) == TK_TUPLETYPE)
```

Comment:
No, we only need `__digestof` when using `digestof` on an abstract type (i.e. not a tuple) with boxed subtypes.